File: packages/python/slop-ai/src/slop/diff.py
"""Recursive diff of two SLOP trees producing JSON Patch operations.

Paths in the generated ops use node IDs for children segments (not array
indices), matching the SLOP patch convention.
"""

from __future__ import annotations

import json
from typing import Any

from .types import PatchOp, SlopNode


def diff_nodes(
    old: SlopNode,
    new: SlopNode,
    base_path: str = "",
) -> list[PatchOp]:
    """Recursively diff *old* and *new* and return patch operations."""
    ops: list[PatchOp] = []

    # --- properties ---
    old_props = old.properties or {}
    new_props = new.properties or {}
    all_keys = set(old_props) | set(new_props)
    for key in sorted(all_keys):
        old_val = old_props.get(key)
        new_val = new_props.get(key)
        if old_val is None and new_val is not None:
            ops.append(PatchOp(op="add", path=f"{base_path}/properties/{key}", value=new_val))
        elif old_val is not None and new_val is None:
            ops.append(PatchOp(op="remove", path=f"{base_path}/properties/{key}"))
        elif _json_ne(old_val, new_val):
            ops.append(PatchOp(op="replace", path=f"{base_path}/properties/{key}", value=new_val))

    # --- affordances (replace entire list if changed) ---
    old_aff = [a.to_dict() for a in old.affordances] if old.affordances else None
    new_aff = [a.to_dict() for a in new.affordances] if new.affordances else None
    if _json_ne(old_aff, new_aff):
        if new_aff is not None:
            op = "replace" if old_aff is not None else "add"
            ops.append(PatchOp(op=op, path=f"{base_path}/affordances", value=new_aff))
        elif old_aff is not None:
            ops.append(PatchOp(op="remove", path=f"{base_path}/affordances"))

    # --- meta (replace entire object if changed) ---
    old_meta = old.meta.to_dict() if old.meta else None
    new_meta = new.meta.to_dict() if new.meta else None
    if _json_ne(old_meta, new_meta):
        if new_meta is not None:
            op = "replace" if old_meta is not None else "add"
            ops.append(PatchOp(op=op, path=f"{base_path}/meta", value=new_meta))
        elif old_meta is not None:
            ops.append(PatchOp(op="remove", path=f"{base_path}/meta"))

    # --- children ---
    old_children = old.children or []
    new_children = new.children or []
    old_map = {c.id: c for c in old_children}
    new_map = {c.id: c for c in new_children}

    # Removed
    for child in old_children:
        if child.id not in new_map:
            ops.append(PatchOp(op="remove", path=f"{base_path}/children/{child.id}"))

    # Added
    for child in new_children:
        if child.id not in old_map:
            ops.append(PatchOp(op="add", path=f"{base_path}/children/{child.id}", value=child.to_dict()))

    # Recursively diff shared children
    for child in new_children:
        old_child = old_map.get(child.id)
        if old_child is not None:
            ops.extend(diff_nodes(old_child, child, f"{base_path}/children/{child.id}"))

    return ops


def _json_ne(a: Any, b: Any) -> bool:
    """Return True if *a* and *b* differ when serialized as JSON."""
    if a is b:
        return False
    if a is None or b is None:
        return a is not b
    return json.dumps(a, sort_keys=True) != json.dumps(b, sort_keys=True)

File: packages/typescript/core/src/diff.ts
import type { SlopNode, PatchOp } from "./types";

/**
 * Recursively diff two SLOP trees and produce JSON Patch ops.
 * Paths use node IDs in children segments (not array indices).
 */
export function diffNodes(
  oldNode: SlopNode,
  newNode: SlopNode,
  basePath: string = ""
): PatchOp[] {
  const ops: PatchOp[] = [];

  // Diff properties key-by-key
  const oldProps = oldNode.properties ?? {};
  const newProps = newNode.properties ?? {};
  const allKeys = new Set([
    ...Object.keys(oldProps),
    ...Object.keys(newProps),
  ]);
  for (const key of allKeys) {
    const oldVal = oldProps[key];
    const newVal = newProps[key];
    if (oldVal === undefined && newVal !== undefined) {
      ops.push({ op: "add", path: `${basePath}/properties/${key}`, value: newVal });
    } else if (oldVal !== undefined && newVal === undefined) {
      ops.push({ op: "remove", path: `${basePath}/properties/${key}` });
    } else if (JSON.stringify(oldVal) !== JSON.stringify(newVal)) {
      ops.push({ op: "replace", path: `${basePath}/properties/${key}`, value: newVal });
    }
  }

  // Diff affordances (replace entire array if changed)
  if (JSON.stringify(oldNode.affordances) !== JSON.stringify(newNode.affordances)) {
    if (newNode.affordances) {
      ops.push({
        op: oldNode.affordances ? "replace" : "add",
        path: `${basePath}/affordances`,
        value: newNode.affordances,
      });
    } else if (oldNode.affordances) {
      ops.push({ op: "remove", path: `${basePath}/affordances` });
    }
  }

  // Diff meta (replace entire object if changed)
  if (JSON.stringify(oldNode.meta) !== JSON.stringify(newNode.meta)) {
    if (newNode.meta) {
      ops.push({
        op: oldNode.meta ? "replace" : "add",
        path: `${basePath}/meta`,
        value: newNode.meta,
      });
    } else if (oldNode.meta) {
      ops.push({ op: "remove", path: `${basePath}/meta` });
    }
  }

  // Diff content_ref (replace entire object if changed)
  if (JSON.stringify(oldNode.content_ref) !== JSON.stringify(newNode.content_ref)) {
    if (newNode.content_ref) {
      ops.push({
        op: oldNode.content_ref ? "replace" : "add",
        path: `${basePath}/content_ref`,
        value: newNode.content_ref,
      });
    } else if (oldNode.content_ref) {
      ops.push({ op: "remove", path: `${basePath}/content_ref` });
    }
  }

  // Diff children
  const oldChildren = oldNode.children ?? [];
  const newChildren = newNode.children ?? [];
  const oldMap = new Map(oldChildren.map((c) => [c.id, c]));
  const newMap = new Map(newChildren.map((c) => [c.id, c]));

  for (const child of oldChildren) {
    if (!newMap.has(child.id)) {
      ops.push({ op: "remove", path: `${basePath}/${child.id}` });
    }
  }

  for (const child of newChildren) {
    if (!oldMap.has(child.id)) {
      ops.push({ op: "add", path: `${basePath}/${child.id}`, value: child });
    }
  }

  for (const child of newChildren) {
    const oldChild = oldMap.get(child.id);
    if (oldChild) {
      ops.push(...diffNodes(oldChild, child, `${basePath}/${child.id}`));
    }
  }

  return ops;
}
